table/tables/tables.go

```diff
@@ -947,59 +945,76 @@ func GetColDefaultValue(ctx sessionctx.Context, col *table.Column, defaultVals [
 
 // AllocHandle implements table.Table AllocHandle interface.
 func (t *tableCommon) AllocHandle(ctx sessionctx.Context) (int64, error) {
-	_, rowID, err := t.Allocator(ctx).Alloc(t.tableID, 1)
+	_, rowID, err := t.Allocator(ctx, autoid.RowIDAllocType).Alloc(t.tableID, 1)
 	if err != nil {
 		return 0, err
 	}
 	if t.meta.ShardRowIDBits > 0 {
 		// Use max record ShardRowIDBits to check overflow.
-		if OverflowShardBits(rowID, t.meta.MaxShardRowIDBits) {
+		if OverflowShardBits(rowID, t.meta.MaxShardRowIDBits, mysql.RowIDBitLength) {
 			// If overflow, the rowID may be duplicated. For examples,
 			// t.meta.ShardRowIDBits = 4
 			// rowID = 0010111111111111111111111111111111111111111111111111111111111111
-			// shard = 01000000000000000000000000000000000000000000000000000000000000000
+			// shard = 0100000000000000000000000000000000000000000000000000000000000000
 			// will be duplicated with:
 			// rowID = 0100111111111111111111111111111111111111111111111111111111111111
 			// shard = 0010000000000000000000000000000000000000000000000000000000000000
 			return 0, autoid.ErrAutoincReadFailed
 		}
 		txnCtx := ctx.GetSessionVars().TxnCtx
 		if txnCtx.Shard == nil {
-			shard := t.calcShard(txnCtx.StartTS)
+			shard := CalcShard(t.meta.ShardRowIDBits, txnCtx.StartTS, mysql.RowIDBitLength)
 			txnCtx.Shard = &shard
 		}
 		rowID |= *txnCtx.Shard
 	}
 	return rowID, nil
 }
 
-// OverflowShardBits checks whether the rowID overflow `1<<(64-shardRowIDBits-1) -1`.
-func OverflowShardBits(rowID int64, shardRowIDBits uint64) bool {
-	mask := (1<<shardRowIDBits - 1) << (64 - shardRowIDBits - 1)
-	return rowID&int64(mask) > 0
+// OverflowShardBits checks whether the recordID overflow `1<<(typeBitsLength-shardRowIDBits-1) -1`.
+func OverflowShardBits(recordID int64, shardRowIDBits uint64, typeBitsLength uint64) bool {
+	mask := (1<<shardRowIDBits - 1) << (typeBitsLength - shardRowIDBits - 1)
+	return recordID&int64(mask) > 0
 }
 
-func (t *tableCommon) calcShard(startTS uint64) int64 {
+// CalcShard calculates the shard prefix by hashing the startTS. Make sure OverflowShardBits is false before calling it.
+func CalcShard(shardRowIDBits uint64, startTS uint64, typeBitsLength uint64) int64 {
 	var buf [8]byte
 	binary.LittleEndian.PutUint64(buf[:], startTS)
 	hashVal := int64(murmur3.Sum32(buf[:]))
-	return (hashVal & (1<<t.meta.ShardRowIDBits - 1)) << (64 - t.meta.ShardRowIDBits - 1)
+	return (hashVal & (1<<shardRowIDBits - 1)) << (typeBitsLength - shardRowIDBits - 1)
 }
 
 // Allocator implements table.Table Allocator interface.
-func (t *tableCommon) Allocator(ctx sessionctx.Context) autoid.Allocator {
+func (t *tableCommon) Allocator(ctx sessionctx.Context, allocType autoid.AllocatorType) autoid.Allocator {
```

Comment:
The implementation shares lots of same codes with `AllAllocators`, can we use the result of `AllAllocators` to avoid repeated code?